stablehlo/reference/ProcessGrid.h

```diff
@@ -0,0 +1,125 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_PROCESSGRID_H_
+#define STABLEHLO_REFERENCE_PROCESSGRID_H_
+
+#include <condition_variable>
+#include <cstdint>
+#include <map>
+#include <mutex>
+#include <utility>
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/reference/Tensor.h"
+
+namespace mlir {
+namespace stablehlo {
+
+// Object to represent StableHLO `process_id`.
+struct ProcessId {
+  /// StableHLO `replica_id`.
+  uint32_t replicaId;
+
+  /// StableHLO `partition_id`.
+  uint32_t partitionId;
+
+  bool operator<(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} <
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
+
+  bool operator==(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} ==
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
+};
+
+// Object to represent StableHLO process group.
+struct ProcessGroup : SmallVector<ProcessId> {};
+
+// Object to represent StableHLO process groups.
+class ProcessGroups : public SmallVector<ProcessGroup> {
+ public:
+  /// Returns an array of process groups.
+  SmallVector<ProcessGroup> find(ProcessId &processId);
+};
+
+/// Class to model a process grid.
+class ProcessGrid {
+ public:
+  /// \name Constructors
+  /// @{
+  ProcessGrid(uint32_t numReplicas, uint32_t numPartitions);
+  /// @}
+
+  /// Cross-replica communication strategy.
+  ProcessGroups crossReplica(SmallVector<SmallVector<uint32_t>> replicaGroups);
+
+  /// Cross-partition communication strategy.
+  ProcessGroups crossPartition(
+      SmallVector<SmallVector<uint32_t>> partitionGroups);
+
+  /// Synchronizes running processes in the `ProcessGroup` and returns the
+  /// collected resources from all partitipating processes in the process group.
+  SmallVector<std::pair<ProcessId, Tensor>> rendezvous(
+      ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
+      const Tensor &operand);
+
+ private:
+  /// The number of replicas the interpreter models.
+  uint32_t numReplicas_;
+
+  /// The number of partitions the interpreter models.
+  uint32_t numPartitions_;
+
+  /// Synchronization primitive used in `rendezvous` function.
+  std::map<std::pair<ProcessGroup, int64_t>, std::mutex> resourceLockMap_;
+
+  /// Synchronization primitive used in `rendezvous` function.
+  std::map<std::pair<ProcessGroup, int64_t>, std::condition_variable>
+      resourceConditionMap_;
+
+  /// Internal mapping of StableHLO `channel_id` to its value.
+  std::map<std::pair<ProcessGroup, int64_t>,
+           SmallVector<std::pair<ProcessId, Tensor>>>
+      channels_;
```

Comment:
```C++
  /// Internal storage used to implement `rendezvous`.
  /// Each call to `rendezvous`, i.e. each combination `processGroup` and
  /// `channelId`, has its own key in the map.
  /// Within the implementation of `rendezvous`, the value corresponding to
  /// this key is gradually populated with tensors arriving from different
  /// processes in the process group.
  std::map<std::pair<ProcessGroup, ChannelId>, RendezvousResult> channels_;

  /// Synchronization primitive used to manage concurrent access to `channels_`.
  std::map<std::pair<ProcessGroup, ChannelId>, std::mutex> channelLocks_;

  /// Synchronization primitive used to manage concurrent access to `channels_`.
  std::map<std::pair<ProcessGroup, ChannelId>, std::condition_variable>
      channelConditions_;
```

Feedback:
  1) It's unclear what's a value of a channel_id, so I elaborated on that.
  2) Dropped the `Map` prefixes, because I don't think much is gained by signaling in the name that this is a map - we can kind of get this already from the signature on the definition site and from the called methods on use sites.
  3) Renamed the `resource` prefix to clearly associate these maps with `channels_`.